pkg/wireguard/agent/agent.go

```diff
@@ -47,18 +52,28 @@ const (
 
 var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
 
+// wireguardClient is an interface to mock wgctrl.Client
+type wireguardClient interface {
+	io.Closer
+	Devices() ([]*wgtypes.Device, error)
+	Device(name string) (*wgtypes.Device, error)
+	ConfigureDevice(name string, cfg wgtypes.Config) error
+}
+
 // Agent needs to be initialized with Init(). In Init(), the Wireguard tunnel
 // device will be created and the proper routes set.  During Init(), existing
 // peer keys are placed into `restoredPubKeys`.  Once RestoreFinished() is
 // called obsolete keys and peers are removed.  UpdatePeer() inserts or updates
 // the public key of peer discovered via the node manager.
 type Agent struct {
 	lock.RWMutex
-	wgClient         *wgctrl.Client
+	wgClient         wireguardClient
+	ipCache          *ipcache.IPCache
 	listenPort       int
 	privKey          wgtypes.Key
-	pubKeyByNodeName map[string]string // nodeName => pubKey
-	restoredPubKeys  map[string]struct{}
+	peerByNodeName   map[string]*peerConfig
+	nodeNameByNodeIP map[string]string
```

Comment:
Reviewer note: Previous discussion as to why this is needed: https://github.com/cilium/cilium/pull/15514#discussion_r608261519